File: src/Tutorials/DifficultyAbility.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.Collections.Generic;
using Microsoft.ML.Probabilistic.Models;
using Microsoft.ML.Probabilistic.Utilities;
using Microsoft.ML.Probabilistic.Distributions;
using Microsoft.ML.Probabilistic.Math;
using Range = Microsoft.ML.Probabilistic.Models.Range;

namespace Microsoft.ML.Probabilistic.Tutorials
{
    [Example("Applications", "Difficulty versus ability in test taking")]
    public class DifficultyAbility
    {
        public void Run()
        {
            InferenceEngine engine = new InferenceEngine();
            if (!(engine.Algorithm is Algorithms.ExpectationPropagation))
            {
                Console.WriteLine("This example only runs with Expectation Propagation");
                return;
            } 
            
            Rand.Restart(0);

            int nQuestions = 100;
            int nSubjects = 40;
            int nChoices = 4;
            Gaussian abilityPrior = new Gaussian(0, 1);
            Gaussian difficultyPrior = new Gaussian(0, 1);
            Gamma discriminationPrior = Gamma.FromMeanAndVariance(1, 0.01);

            double[] trueAbility, trueDifficulty, trueDiscrimination;
            int[] trueTrueAnswer;
            int[][] data = Sample(
                nSubjects,
                nQuestions,
                nChoices,
                abilityPrior,
                difficultyPrior,
                discriminationPrior,
                out trueAbility,
                out trueDifficulty,
                out trueDiscrimination,
                out trueTrueAnswer);
            Range question = new Range(nQuestions).Named("question");
            Range subject = new Range(nSubjects).Named("subject");
            Range choice = new Range(nChoices).Named("choice");
            var response = Variable.Array(Variable.Array<int>(question), subject).Named("response");
            response.ObservedValue = data;

            var ability = Variable.Array<double>(subject).Named("ability");
            ability[subject] = Variable.Random(abilityPrior).ForEach(subject);
            var difficulty = Variable.Array<double>(question).Named("difficulty");
            difficulty[question] = Variable.Random(difficultyPrior).ForEach(question);
            var discrimination = Variable.Array<double>(question).Named("discrimination");
            discrimination[question] = Variable.Random(discriminationPrior).ForEach(question);
            var trueAnswer = Variable.Array<int>(question).Named("trueAnswer");
            trueAnswer[question] = Variable.DiscreteUniform(choice).ForEach(question);

            using (Variable.ForEach(subject))
            {
                using (Variable.ForEach(question))
                {
                    var advantage = (ability[subject] - difficulty[question]).Named("advantage");
                    var advantageNoisy = Variable.GaussianFromMeanAndPrecision(advantage, discrimination[question]).Named("advantageNoisy");
                    var correct = (advantageNoisy > 0).Named("correct");
                    using (Variable.If(correct))
                    {
                        response[subject][question] = trueAnswer[question];
                    }

                    using (Variable.IfNot(correct))
                    {
                        response[subject][question] = Variable.DiscreteUniform(choice);
                    }
                }
            }

            engine.NumberOfIterations = 5;
            subject.AddAttribute(new Models.Attributes.Sequential());  // needed to get stable convergence
            question.AddAttribute(new Models.Attributes.Sequential());  // needed to get stable convergence
            bool doMajorityVoting = false; // set this to 'true' to do majority voting
            if (doMajorityVoting)
            {
                ability.ObservedValue = Util.ArrayInit(nSubjects, i => 0.0);
                difficulty.ObservedValue = Util.ArrayInit(nQuestions, i => 0.0);
                discrimination.ObservedValue = Util.ArrayInit(nQuestions, i => 1.0);
            }

            var trueAnswerPosterior = engine.Infer<IList<Discrete>>(trueAnswer);
            int numCorrect = 0;
            for (int q = 0; q < nQuestions; q++)
            {
                int bestGuess = trueAnswerPosterior[q].GetMode();
                if (bestGuess == trueTrueAnswer[q])
                {
                    numCorrect++;
                }
            }

            double pctCorrect = 100.0 * numCorrect / nQuestions;
            Console.WriteLine("{0}% TrueAnswers correct", pctCorrect.ToString("f0"));
            var difficultyPosterior = engine.Infer<IList<Gaussian>>(difficulty);
            for (int q = 0; q < System.Math.Min(nQuestions, 4); q++)
            {
                Console.WriteLine("difficulty[{0}] = {1} (sampled from {2})", q, difficultyPosterior[q], trueDifficulty[q].ToString("g2"));
            }

            var discriminationPosterior = engine.Infer<IList<Gamma>>(discrimination);
            for (int q = 0; q < System.Math.Min(nQuestions, 4); q++)
            {
                Console.WriteLine("discrimination[{0}] = {1} (sampled from {2})", q, discriminationPosterior[q], trueDiscrimination[q].ToString("g2"));
            }

            var abilityPosterior = engine.Infer<IList<Gaussian>>(ability);
            for (int s = 0; s < System.Math.Min(nSubjects, 4); s++)
            {
                Console.WriteLine("ability[{0}] = {1} (sampled from {2})", s, abilityPosterior[s], trueAbility[s].ToString("g2"));
            }
        }

        public int[][] Sample(
            int nSubjects,
            int nQuestions,
            int nChoices,
            Gaussian abilityPrior,
            Gaussian difficultyPrior,
            Gamma discriminationPrior,
            out double[] ability,
            out double[] difficulty,
            out double[] discrimination,
            out int[] trueAnswer)
        {
            ability = Util.ArrayInit(nSubjects, s => abilityPrior.Sample());
            difficulty = Util.ArrayInit(nQuestions, q => difficultyPrior.Sample());
            discrimination = Util.ArrayInit(nQuestions, q => discriminationPrior.Sample());
            trueAnswer = Util.ArrayInit(nQuestions, q => Rand.Int(nChoices));
            int[][] response = new int[nSubjects][];
            for (int s = 0; s < nSubjects; s++)
            {
                response[s] = new int[nQuestions];
                for (int q = 0; q < nQuestions; q++)
                {
                    double advantage = ability[s] - difficulty[q];
                    double noise = Gaussian.Sample(0, discrimination[q]);
                    bool correct = (advantage > noise);
                    if (correct)
                    {
                        response[s][q] = trueAnswer[q];
                    }
                    else
                    {
                        response[s][q] = Rand.Int(nChoices);
                    }
                }
            }

            return response;
        }
    }
}

File: test/TestFSharp/DifficultyAbility.fs
namespace TestFSharp

open System.Collections.Generic
open Microsoft.ML.Probabilistic
open Microsoft.ML.Probabilistic.FSharp
open Microsoft.ML.Probabilistic.Models
open Microsoft.ML.Probabilistic.Utilities
open Microsoft.ML.Probabilistic.Distributions
open Microsoft.ML.Probabilistic.Math

module DifficultyAbility =
    let main() = 
        Rand.Restart(0);
        let nQuestions = 100
        let nSubjects = 40
        let nChoices = 4
        let abilityPrior = Gaussian(0.0, 1.0)
        let difficultyPrior = Gaussian(0.0, 1.0)
        let discriminationPrior = Gamma.FromMeanAndVariance(1.0, 0.01)

        let Sample(nSubjects:int,nQuestions:int,nChoices:int,abilityPrior:Gaussian,difficultyPrior:Gaussian,discriminationPrior:Gamma)=
          let ability= Util.ArrayInit( nSubjects, (fun _-> abilityPrior.Sample()))
          let difficulty = Util.ArrayInit(nQuestions, (fun _ -> difficultyPrior.Sample()))
          let discrimination = Util.ArrayInit(nQuestions, (fun _ -> discriminationPrior.Sample()))
          let trueAnswer = Util.ArrayInit(nQuestions, (fun _ -> Rand.Int(nChoices)))
          let response:int[][] = Array.zeroCreate nSubjects 
          for s in 0..(nSubjects-1) do
            response.[s] <- Array.zeroCreate nQuestions
            for q in 0..(nQuestions-1) do
              let advantage = ability.[s] - difficulty.[q]
              let noise = Gaussian.Sample(0.0, discrimination.[q])
              let correct = (advantage > noise)
              if (correct) then
                  response.[s].[q] <- trueAnswer.[q]
              else
                  response.[s].[q] <- Rand.Int(nChoices)
          (response, ability,difficulty,discrimination,trueAnswer)


        let data,trueAbility,trueDifficulty,trueDiscrimination,trueTrueAnswer = Sample(nSubjects,nQuestions,nChoices,abilityPrior,difficultyPrior,discriminationPrior)

        let question = Range(nQuestions).Named("question")
        let subject = Range(nSubjects).Named("subject")
        let choice = Range(nChoices).Named("choice")

        //let response = Variable.Array(Variable.Array<int>(question), subject).Named("response")
        let response = Variable.Array<VariableArray<int>, int [][]>(Variable.Array<int>(question), subject).Named("response")

        response.ObservedValue <- data

        let ability = Variable.Array<double>(subject).Named("ability")
        Variable.ForeachBlock subject ( fun s ->  ability.[s] <- Variable.Random(abilityPrior) )
        let difficulty = Variable.Array<double>(question).Named("difficulty")
        Variable.ForeachBlock question ( fun q ->  difficulty.[q] <- Variable.Random(difficultyPrior) )
        let discrimination = Variable.Array<double>(question).Named("discrimination")
        Variable.ForeachBlock question ( fun q ->  discrimination.[q] <- Variable.Random(discriminationPrior) )
        let trueAnswer = Variable.Array<int>(question).Named("trueAnswer")
        Variable.ForeachBlock question ( fun q ->  trueAnswer.[q] <- Variable.DiscreteUniform(choice) )

        Variable.ForeachBlock subject (fun s -> 
          Variable.ForeachBlock question (fun q -> 
            let advantage = (ability.[s] - difficulty.[q]).Named("advantage")
            let advantageNoisy = Variable.GaussianFromMeanAndPrecision(advantage, discrimination.[q]).Named("advantageNoisy")
            let correct = (advantageNoisy >> 0.0).Named("correct")
            Variable.IfBlock correct (fun _->response.[s].[q] <- trueAnswer.[q]) (fun _->response.[s].[q] <- Variable.DiscreteUniform(choice))
            ()
          )
        )

        let engine = InferenceEngine()
        engine.NumberOfIterations <- 5
        subject.AddAttribute(Models.Attributes.Sequential())
        question.AddAttribute(Models.Attributes.Sequential())
        let doMajorityVoting = false; // set this to 'true' to do majority voting
        if doMajorityVoting then
            ability.ObservedValue <- Util.ArrayInit(nSubjects, (fun i -> 0.0))
            difficulty.ObservedValue <- Util.ArrayInit(nQuestions, (fun i -> 0.0))
            discrimination.ObservedValue <- Util.ArrayInit(nQuestions, (fun i -> 0.0))

        let trueAnswerPosterior = engine.Infer<IReadOnlyList<Discrete>>(trueAnswer)

        let mutable numCorrect = 0
        for q in 0..(nQuestions-1) do
            let bestGuess = trueAnswerPosterior.[q].GetMode()
            if (bestGuess = trueTrueAnswer.[q]) then
              numCorrect<-numCorrect+1

        let pctCorrect:float = 100.0 * (float numCorrect) / (float nQuestions)
        printfn "%f TrueAnswers correct" pctCorrect

        let difficultyPosterior = engine.Infer<IReadOnlyList<Gaussian>>(difficulty)


        for q in 0..(System.Math.Min(nQuestions, 4)-1) do
          printfn "difficulty[%i] = %A (sampled from %f)"  q  difficultyPosterior.[q] trueDifficulty.[q]

        let discriminationPosterior = engine.Infer<IReadOnlyList<Gamma>>(discrimination)
        for q in 0..(System.Math.Min(nQuestions, 4)-1) do
          printfn "discrimination[%i] = %A (sampled from %f)"  q  discriminationPosterior.[q] trueDiscrimination.[q]

        let abilityPosterior = engine.Infer<IReadOnlyList<Gaussian>>(ability)
        for s in 0..(System.Math.Min(nQuestions, 4)-1) do
          printfn "ability[%i] = %A (sampled from %f)"  s  abilityPosterior.[s] trueAbility.[s]

